Bugfix: skip NULL fields when reading from MySQL

chenzhuoyu opened the pull request:
In `drizzle_state_binary_field_read()` at `field.cc:401`

`NULL` fields have no bytes but bit-masked by `con->result->null_bitmap`, we should skip them.

For example, given table:
```sql
CREATE TABLE `test_table` (
  a VARCHAR(10) DEFAULT NULL,
  b VARCHAR(10) DEFAULT NULL,
  c INT(10) DEFAULT NULL,
  d VARCHAR(10) DEFAULT NULL
) ENGINE=InnoDB;
```

and row

| a     | b      | c   | d     |
| ----- | ------ | --- | ----- |
| test1 | `NULL` | 123 | test2 |

Since `NULL` fields are not presented in the response body, if we try to read field `b` from the response, it's not a valid operation and may cause serious problems.

This patch fixes this problem by skipping `NULL` fields when reading from MySQL's response.

chenzhuoyu committed:
Bugfix: skip NULL fields when reading from MySQL

andreas-bok-sociomantic commented on libdrizzle/statement.cc:
```diff
@@ -392,7 +392,6 @@ drizzle_return_t drizzle_stmt_fetch(drizzle_stmt_st *stmt)
   drizzle_return_t ret= DRIZZLE_RETURN_OK;
   uint16_t column_counter= 0, current_column= 0;
   drizzle_row_t row;
-  drizzle_column_st *column;
```
Do we need to move the declaration of the `drizzle_column_st ` into the for loop?

chenzhuoyu replied:
Not necessarily, but it looks a bit nicer

andreas-bok-sociomantic commented on libdrizzle/field.cc:
```diff
@@ -402,9 +402,17 @@ drizzle_return_t drizzle_state_binary_field_read(drizzle_st *con)
 
-  switch(con->result->column_buffer[con->result->field_current].type)
+  /* Skip NULL fields */
+  while (con->result->null_bitmap[(con->result->field_current_read + 2) / 8] & (1 << ((con->result->field_current_read + 2) % 8)))
```
Just understand. Is the added offset of 2 in `con->result->field_current_read + 2` a consequence of the cautionary note in the MySQL internals doc for NULL-Bitmap?, cf. https://dev.mysql.com/doc/internals/en/null-bitmap.html

andreas-bok-sociomantic replied:
Please format as:

```c++
  while (con->result->null_bitmap[(con->result->field_current_read + 2) / 8] &
         (1 << ((con->result->field_current_read + 2) % 8)))
```

chenzhuoyu replied:
1. Yes, and I've captured the actual packet on wire using WireShark, confirmed that 2-bits offset is actually there, but in fact I have no idea why, and I found nothing about this magic number in that manual either :(  Maybe they want to maintain protocol compatibility ?

2. OK, I'll do it tomorrow :)

andreas-bok-sociomantic reviewed (commented):
Hi @chenzhuoyu, thanks for your contribution :)
LGTM just have a few comments. 
Just to clarify. Is the changes based on the MySQL doc for NULL-bitmap?  https://dev.mysql.com/doc/internals/en/null-bitmap.html. 

Secondly, we haven't encountered the problems with would this bug could have caused. We read data with NULL values which is parsed by the current implementation. 
Is it possible to make a test case which would reproduce the bug?

chenzhuoyu commented:
We found this bug by running unittests in our project, for some reason I cannot provide the source code, but I will construct a seperate test program to reproduce this problem :)

chenzhuoyu commented:
.....Oops I clicked the wrong button OMG .....

chenzhuoyu commented:
Here I constructed a simple program to perfrom the test:
[Test case, on my Gist](https://gist.github.com/chenzhuoyu/70554377c5e82a78bae4becc31e0873b)

g++ version:
```
g++ (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Linux version:
```
Linux 4.4.0-31-generic #50-Ubuntu SMP Wed Jul 13 00:07:12 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

MySQL version:
```
mysql  Ver 14.14 Distrib 5.7.17, for Linux (x86_64) using  EditLine wrapper
```

libdrizzle-redux configure command:
```bash
./configure --prefix=/usr/local --enable-shared=no --enable-static=yes
```

libdrizzle-redux build command:
```bash
make -j24
make install
```

Test case build command:
```bash
g++ -g -Wall -pthread -o test test.cpp -ldrizzle-redux -lssl -lcrypto -lz
```

Table creation SQL:
```sql
CREATE TABLE `test_table` (
    `id`        int(10)         NOT NULL AUTO_INCREMENT,
    `word`      varchar(255)    NOT NULL,
    `time`      varchar(16)     NOT NULL,
    `uid`       varchar(32)     DEFAULT NULL,
    `question`  varchar(255)    DEFAULT NULL,
    `answer`    varchar(255)    DEFAULT NULL,
    `is_valid`  int(1)          NOT NULL DEFAULT 1,
    `source`    varchar(32)     DEFAULT NULL,
    `type`      int(2)          DEFAULT NULL,
    `uuid`      char(36)        DEFAULT NULL,
    PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8
```

with content:
```sql
INSERT INTO `test_table` (`word`, `time`, `question`, `answer`, `is_valid`, `source`, `uuid`)
VALUES ('test word', '1492672507', 'test question', 'test answer', 0, 'x', 'd7f15abb-a482-43be-99c2-c175408c1bd9');
```

MySQL client result:
```
mysql> SELECT * FROM test_table;
+----+-----------+------------+------+---------------+-------------+----------+--------+------+--------------------------------------+
| id | word      | time       | uid  | question      | answer      | is_valid | source | type | uuid                                 |
+----+-----------+------------+------+---------------+-------------+----------+--------+------+--------------------------------------+
|  1 | test word | 1492672507 | NULL | test question | test answer |        0 | x      | NULL | d7f15abb-a482-43be-99c2-c175408c1bd9 |
+----+-----------+------------+------+---------------+-------------+----------+--------+------+--------------------------------------+
1 row in set (0.00 sec)
```

My result of this test case:
```
$ ./test
execute
buffer
** FATAL : [9] drizzle_state_packet_read:bad packet number:14:0
```

Also the capture file is attached here:
[capture.zip](https://github.com/sociomantic-tsunami/libdrizzle-redux/files/960839/capture.zip)

chenzhuoyu commented:
@andreas-bok-sociomantic hope this helps :)

chenzhuoyu commented:
@andrea-imparato-sociomantic I've updated the code format

andreas-bok-sociomantic commented:
Great with the test case for the bug. 
The PR looks good to me, so just a matter of me running the test case on my computer. 
Get back to you soon

andreas-bok-sociomantic committed:
Add unittest for prepared statement with buffering

Covers the bug reported by @chenzhuoyu (#153)
where read fields with null values were
not skipped

andreas-bok-sociomantic commented:
Hi @chenzhuoyu, ran the testcase you provided, And indeed it fails when buffering a prepared statement query. 
I created a unittest from your example and pushed it to this branch https://github.com/andreas-bok-sociomantic/libdrizzle-redux/tree/skip-null-fields
You are welcome to add the [commit](https://github.com/andreas-bok-sociomantic/libdrizzle-redux/commit/909f8090178564658f4bb14cbc75b677c20d58f9) with the added test case to this PR. 
Otherwise I will just make a new PR with that addition. 

Thanks again for your great work :)

andreas-bok-sociomantic commented:
ping @chenzhuoyu

chenzhuoyu commented:
I've cherry picked the test-case commit from your repository @andreas-bok-sociomantic 
Sorry for the late reply, I'v been on vacation these days and did not have chance to read your email before.

andreas-bok-sociomantic commented:
Cool, no worries. I will just take one more look. But it should ready to merge :)

andreas-bok-sociomantic commented on libdrizzle/field.cc:
```diff
@@ -402,9 +402,18 @@ drizzle_return_t drizzle_state_binary_field_read(drizzle_st *con)
+  switch(con->result->column_buffer[con->result->field_current_read].type)
   {
     case DRIZZLE_COLUMN_TYPE_NULL:
+      /* This should not happen since NULL fields are masked by `null_bitmap`, but cover it anyway */
```
A colleague, @daniel-zullo-sociomantic , noted that if the unlikely case should happen and this case is hit, `field_current_read` would be incorrectly updated twice. 
How do you feel about this instead, which would indeed make it impossible to parse a field with a null value:

```cpp 
  /* Skip NULL fields */
  while (con->result->null_bitmap[(con->result->field_current_read + 2) / 8] &
         (1 << ((con->result->field_current_read + 2) % 8)))
  {
    con->result->field_current_read++;
  }

  drizzle_column_type_t column_type =
    con->result->column_buffer[con->result->field_current_read].type;

  assert(column_type !=  DRIZZLE_COLUMN_TYPE_NULL);

  switch(column_type)
  {
    case DRIZZLE_COLUMN_TYPE_TINY:
      con->result->field_size= 1;
      break;
  // ...
  }
```

andreas-bok-sociomantic replied:
This might be better since it won't halt the program and will allow the user to handle the error and get an error message 

```cpp
drizzle_return_t drizzle_state_binary_field_read(drizzle_st *con)
...
  /* Skip NULL fields */
  while (con->result->null_bitmap[(con->result->field_current_read + 2) / 8] &
         (1 << ((con->result->field_current_read + 2) % 8)))
  {
    con->result->field_current_read++;
  }

  drizzle_column_type_t column_type =
    con->result->column_buffer[con->result->field_current_read].type;

  switch(column_type)
  {
    case DRIZZLE_COLUMN_TYPE_NULL:
      /* This should not happen since NULL fields are masked by `null_bitmap`,
       * but cover it anyway */
      drizzle_set_error(con, __func__,
        "Unexpected column type: DRIZZLE_COLUMN_TYPE_NULL");
      con->result->field_size= 0;
      return DRIZZLE_RETURN_UNEXPECTED_DATA;
      break;
    case DRIZZLE_COLUMN_TYPE_TINY:
...
```

andreas-bok-sociomantic commented:
ping @chenzhuoyu

chenzhuoyu commented:
Sorry for the late reply, but I can't figure out in which case `field_current_read` could be updated twice :(

Actually I don't even think `column_type` could be `DRIZZLE_COLUMN_TYPE_NULL` ever.

I've gone through the source (by `fgrep -nR 'DRIZZLE_COLUMN_TYPE_NULL' .`), and found there were only 2 places that sets something to `DRIZZLE_COLUMN_TYPE_NULL`, which are:

1. The first one is for setting query parameter types in prepared statements, thus have nothing to do with result column types.

```cpp
/* see statement_param.cc:134 in "drizzle_stmt_set_null" */
/* 132 */ drizzle_return_t drizzle_stmt_set_null(drizzle_stmt_st *stmt, uint16_t param_num)
/* 133 */ {
/* 134 */   return drizzle_stmt_set_param(stmt, param_num, DRIZZLE_COLUMN_TYPE_NULL, NULL, 0, false);
/* 135 */ }
```

which invokes `drizzle_stmt_set_param`, at line `58` it sets the query parameter type

```cpp
/* see statement_param.cc:58 in "drizzle_stmt_set_param" */
/* 46 */ drizzle_return_t drizzle_stmt_set_param(drizzle_stmt_st *stmt, uint16_t param_num, drizzle_column_type_t type, const void *data, size_t length, bool is_unsigned)
/* 47 */ {
/* 48 */   if ((stmt == NULL) || (param_num >= stmt->param_count))
/* 49 */   {
/* 50 */     return DRIZZLE_RETURN_INVALID_ARGUMENT;
/* 51 */   }
/* 52 */   if (stmt->state < DRIZZLE_STMT_PREPARED)
/* 53 */   {
/* 54 */     drizzle_set_error(stmt->con, __func__, "stmt object has not been prepared");
/* 55 */     return DRIZZLE_RETURN_STMT_ERROR;
/* 56 */   }
/* 57 */ 
/* 58 */   stmt->query_params[param_num].type= type;
/* 59 */   stmt->query_params[param_num].data= (void*)data;
/* 60 */   stmt->query_params[param_num].length= length;
/* 61 */   stmt->query_params[param_num].options.is_unsigned= is_unsigned;
/* 62 */   stmt->query_params[param_num].is_bound= true;
/* 63 */ 
/* 64 */   return DRIZZLE_RETURN_OK;
/* 65 */ }
```

2. The second one happens **AFTER** the row been buffered, also have nothing to do with result column types.

```cpp
/* see statement.cc:435 in "drizzle_stmt_fetch" */
/* 410 */ /* ... */
/* 411 */ if (stmt->execute_result->options & DRIZZLE_RESULT_BUFFER_ROW)
/* 412 */ {
/* 413 */   row= drizzle_row_next(stmt->execute_result);
/* 414 */ }
/* 415 */ else
/* 416 */ {
/* 417 */   row= drizzle_row_buffer(stmt->execute_result, &ret);
/* 418 */ }
/* 419 */ 
/* 420 */ if (row == NULL)
/* 421 */ {
/* 422 */   return DRIZZLE_RETURN_ROW_END;
/* 423 */ }
/* 424 */ 
/* 425 */ for (column_counter = 0; column_counter < stmt->execute_result->column_count; column_counter++)
/* 426 */ {
/* 427 */   drizzle_bind_st *param= &stmt->result_params[column_counter];
/* 428 */   drizzle_column_st *column= drizzle_column_index(stmt->execute_result, column_counter);
/* 429 */ 
/* 430 */   /* if this row is null in the result bitmap, skip first 2 bits */
/* 431 */   if (stmt->execute_result->null_bitmap[(column_counter+2)/8] & (1 << ((column_counter+2) % 8)))
/* 432 */   {
/* 433 */     param->options.is_null= true;
/* 434 */     param->length= 0;
/* 435 */     param->type= DRIZZLE_COLUMN_TYPE_NULL;
/* 436 */   }
/* 437 */   else
/* 438 */ /* ... */
```

At line `435` it sets `param->type` to `DRIZZLE_COLUMN_TYPE_NULL`, but this happens **AFTER** both `drizzle_row_next` and `drizzle_row_buffer` (see `statement.cc:413 and :417 in "drizzle_stmt_fetch"`)

But anyway, I agree with you that adding an error message in that "case" statement is a good idea, although I don't think it might ever happens [Facepalm]

chenzhuoyu commented:
@andreas-bok-sociomantic

andreas-bok-sociomantic commented:
> But anyway, I agree with you that adding an error message in that "case" statement is a good idea, although I don't think it might ever happens [Facepalm]

Thanks for looking into it again. I suggest based on your comments, that covering the case (which in fact cannot happen) and setting an error message is sufficient. If not else to have the switch handle all 
`drizzle_column_type_t` enum values. 

With that change I would be happy to merge

```cpp
     case DRIZZLE_COLUMN_TYPE_NULL:
      /* This should not happen since NULL fields are masked by `null_bitmap`,
       * but cover it anyway */
      drizzle_set_error(con, __func__,
        "Unexpected column type: DRIZZLE_COLUMN_TYPE_NULL");
      con->result->field_size= 0;
      return DRIZZLE_RETURN_UNEXPECTED_DATA;
      break;
```

chenzhuoyu commented:
@andreas-bok-sociomantic I've added these changes

andreas-bok-sociomantic commented:
Cool thanks
